File: (结构)组合模式.v1/golang/gpt/index.组合.go
package main

import "fmt"

// Component 提供了组合中对象的共享接口，既可以代表单个对象也可以代表组合对象
type Component interface {
	Perform() string
}

// Leaf 实现了 Component 接口，并代表组合中的单个对象
type Leaf struct {
	name string
}

func (l *Leaf) Perform() string {
	return l.name
}

// Composite 实现了 Component 接口，并代表一个由子部件构成的组合对象
type Composite struct {
	children []Component
	name     string
}

func (c *Composite) Add(child Component) {
	c.children = append(c.children, child)
}

func (c *Composite) Perform() string {
	result := fmt.Sprintf("Composite [%s]\n", c.name)
	for _, child := range c.children {
		result += child.Perform() + "\n"
	}

	// 移除末尾的换行符
	result = result[:len(result)-1]
	return result
}

/**
在Go语言（又称为Golang）中，组合模式不是通过继承实现的，
而是通过嵌入结构体（embedding struct）的方式来实现的。
组合模式是一种结构型设计模式，它可以让你将对象组合成树形结构以表示部分-整体的层次结构。
组合模式通过让单个对象和组合对象共享相同的接口来实现这一点。
*/
func main() {
	// 创建树形结构
	root := &Composite{name: "Root"}

	leaf1 := &Leaf{name: "Leaf 1"}
	leaf2 := &Leaf{name: "Leaf 2"}

	composite1 := &Composite{name: "Composite 1"}
	composite2 := &Composite{name: "Composite 2"}

	root.Add(leaf1)
	root.Add(composite1)

	composite1.Add(leaf2)
	composite1.Add(composite2)

	composite2.Add(&Leaf{name: "Leaf 3"})

	// 执行操作
	fmt.Println(root.Perform())
}

File: (结构)组合模式.v2/php/gpt/index.组合.php
<?php
namespace php\gpt\combine;
interface Component {
    public function operation();
}
class Leaf implements Component {
    private $name;

    public function __construct($name) {
        $this->name = $name;
    }

    public function operation() {
        echo "Leaf " . $this->name . " is operated.\n";
    }
}
class Composite implements Component {
    private $children = [];
    private $name;

    public function __construct($name) {
        $this->name = $name;
    }

    public function add(Component $component) {
        $this->children[] = $component;
    }

    public function remove(Component $component) {
        $this->children = array_filter($this->children, function($child) use ($component) {
            return $child !== $component;
        });
    }

    public function operation() {
        echo "Composite " . $this->name . " operates on its children:\n";
        foreach ($this->children as $child) {
            $child->operation();
        }
    }
}
// 创建叶子节点
$leaf1 = new Leaf('1');
$leaf2 = new Leaf('2');
$leaf3 = new Leaf('3');

// 创建复合节点，并添加叶子节点
$composite = new Composite('Comp1');
$composite->add($leaf1);
$composite->add($leaf2);
$composite->add($leaf3);

// 创建更高一级的复合节点
$rootComposite = new Composite('RootComp');
$rootComposite->add($composite);

// 执行操作
$rootComposite->operation();

/**
    Composite RootComp operates on its children:
    Composite Comp1 operates on its children:
    Leaf 1 is operated.
    Leaf 2 is operated.
    Leaf 3 is operated.
 */